MPR#7640: reimplementation of Unix.execvpe

xavierleroy opened the pull request:
Unlike its siblings `execve()` and `execvp()`, the [`execvpe()`](https://linux.die.net/man/3/execvpe) process-launching function is not part of POSIX and is not provided by the C standard libraries of several systems we care about (MacOS, FreeBSD, etc).

Before OCaml 4.05, `Unix.execvpe` was implementing by changing the global variable `environ` and calling `execvp()`.  But this doesn't work on Cygwin [(MPR#6903)](https://caml.inria.fr/mantis/view.php?id=6903) and has the bizarre effect that the executable is searched in the "new PATH" (the one from the environment given as parameter to `Unix.execvpe`) and not in the current PATH.  Some users came to rely on this bizarre effect, see initial discussion in [MPR#7640](https://caml.inria.fr/mantis/view.php?id=7640), but were talked out of it.

In OCaml 4.05, `Unix.execvpe` was reimplemented by first searching in the PATH, using `caml_search_exe_in_path`, then calling `execve()`.  This is portable and searches the correct PATH.  However, as discussed in the second part of  [MPR#7640](https://caml.inria.fr/mantis/view.php?id=7640), the behavior is not quite consistent with that of `execvp()`.  In particular, owing to the use of `caml_search_exe_in_path`, the current directory is always searched (last, but always) for the given executable name, even if the PATH does not contain the current directory.  This is a security risk.

The present pull request reimplements `Unix.execvpe` once more.
* On platforms that provide a C implementation of `execvpe()`, we just call into it.  This includes Linux, OpenBSD, native Win32 (MSVC and MINGW ports), and Cygwin.
* On other platforms, we use our own implementation, written in OCaml and searching the PATH itself.  The implementation is patterned after that of Glibc [(C source here)](https://sourceware.org/git/?p=glibc.git;a=blob_plain;f=posix/execvpe.c;hb=HEAD) and therefore should satisfy the comments from [MPR#7640](https://caml.inria.fr/mantis/view.php?id=7640), including the more pedantic ones.

A test was added to the test suite.  It runs only if our implementation is selected, because we're not testing other people's implementations of `execvpe()`, and because that of Cygwin behaves somewhat bizarrely.  However, the results of the test suite were validated against the Linux and OpenBSD implementations of `execvpe()`.

xavierleroy committed:
MPR#7640: revised implementation of Unix.execvpe

Use the system-provided execvpe() if possible.  Otherwise, use
a serious reimplementation written in OCaml and patterned after
the Glibc execvpe() implementation.

Added a test in tests/lib-unix/unix-execvpe.

Don't test Unix.execvpe if we are using the system-provided implementation.
The execvpe() functions provided by Win32 and Cygwin aren't quite to
our specs.  At any rate, the test is there to find bugs in our
implementation of execvpe(), not in others's.

aalekseyev commented:
Thanks, this is now much closer to what we have in core and I'm making a change to core to reduce the differences even further.

avsm commented:
I couldn't spot any behavioural anomalies upon reading the diff.  My only query is why we need to maintain a C binding at all anymore, given the existence of the very similar OCaml implementation.  Could we just use the OCaml version on all platforms and reduce our exposure to subtly varying C library code a little bit?

xavierleroy commented:
Thanks for the reviews.  About using the OCaml version on all platforms, I thought about it but got cold feet, because:
* The Win32 ports (MSVC, Mingw) probably do things their own way.  E.g. the search path may come from the PATH environment variable or may come from the registry.
* Likewise the Cygwin implementation is probably special.
* Even on honest Unix systems, there is a risk that our `execvpe` algorithm differs slightly from the algorithm used to implement `execvp` in the C standard library. I feel that consistency between the various `exec` functions on the same platform trumps consistency between platforms.

ivg commented on otherlibs/unix/unix.ml:
```diff
@@ -203,7 +203,68 @@ type wait_flag =
+        (* Other errors, e.g. E2BIG, are fatal and abort the search. *)
+        | _ ->
+            raise exn in
+  if not (Filename.is_implicit name) then
```
The `Filename.is_implicit name` will be false only for file names which start with `/`, `./`, or `../`, though POSIX [prescribes][1] to omit the search when a filename contains a slash (at any position):

>  If the pathname being sought contains a slash, the search through the path prefixes shall not be performed.

E.g., `Filename.is_implicit "bin/ls"` is `true`.

[1]: http://pubs.opengroup.org/onlinepubs/009695399/basedefs/xbd_chap08.html

xavierleroy replied:
Well spotted, thanks.  Fixed in commit [execvpe 79bcc03]

ivg commented on otherlibs/unix/unix.ml:
```diff
@@ -203,7 +203,68 @@ type wait_flag =
+      let argc = Array.length args in
+      (* Drop the original args.(0) if it is there *)
+      let new_args = Array.append
+        [| "/bin/sh"; file |]
```
shouldn't we at least try to call `confstr(_CS_PATH)` or it will complicate things too much? It is POSIX.1-2001, so I would expect this function to be available on most of the platforms, if not all.

xavierleroy replied:
There is no OCaml binding for `confstr` so far, and quite frankly I don't miss it.  The OpenBSD implementation of `execvpe` doesn't care either, using
```
#define	_PATH_BSHELL	"/bin/sh"
```
as the shell to execute.

xavierleroy committed:
MPR#7640: revised implementation of Unix.execvpe, continued

Don't search in path if filename contains a '/', even if it is an "implicit" filename such as "foo/bar".

alainfrisch commented on otherlibs/unix/unix.ml:
```diff
@@ -203,7 +203,68 @@ type wait_flag =
+    exec name
+  else
+    (* Split path into elements and search in these elements *)
+    (try unsafe_getenv "PATH" with Not_found -> "/bin:/usr/bin")
```
Does this default PATH corresponds to something standard?

xavierleroy replied:
It is what glibc uses, so it's standard enough for me.
```
./sysdeps/unix/confstr.h:#define	CS_PATH	"/bin:/usr/bin"
```
If it were me, `execvp*` would fail if the PATH environment variable is unset.

aalekseyev replied:
glibc also adds the current directory by default. Let's say this is an intersection between what glibc and openbsd use. :)

gasche commented:
@xavierleroy, do I correctly understand that you think this should be merged in 4.06, to close the behavioral gap? Do @aalekseyev and @avsm's reviews give you enough confidence in the change for this?

xavierleroy commented:
Yes I think this must go in 4.06 because the 4.05 implementation of `Unix.execvpe` is clunky and can be viewed as a security risk.

Yes I feel confident about the code based on the comments from  @aalekseyev and @avsm and especially on the review by @ivg.

xavierleroy committed:
MPR#7640: update Changes

xavierleroy commented:
Cherry-picked as [4.06 e28cc0a]

avsm commented:
The only slight ancillary risk I can see for 4.06 from this patch is that `_GNU_SOURCE` is now defined in the bigger unix stubs file.  I'm testing it on platforms where that might cause some weird knockon effect like the embedded libcs like uclibc, but I've not spotted any problems so far.
